File: src/main/java/core/procedures/vectors/MakeVector.java
package core.procedures.vectors;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.Type;
import core.scm.MutableVector;

public final class MakeVector extends AFn {

  public MakeVector() {
    super(new FnArgsBuilder().min(1).max(2).mandatory(new Class[]{Type.ExactNonNegativeInteger.class}).build());
  }

  @Override
  public String getName() {
    return "make-vector";
  }

  @Override
  public Object apply(Object... args) {
    Long s = ((Number)args[0]).longValue();
    Object init = null;
    if (args.length == 2) {
      init = args[1];
    }
    return new MutableVector(s.intValue(), init);
  }
}

File: src/main/java/core/procedures/vectors/MakeVector.kt
package core.procedures.vectors

import core.procedures.AFn
import core.procedures.Arity.Range
import core.scm.MutableVector
import core.scm.Type

class MakeVector : AFn<Any?, MutableVector>(name = "make-vector", isPure = true, arity = Range(1, 2),
                       mandatoryArgsTypes = arrayOf(Type.ExactNonNegativeInteger::class.java)) {

    override operator fun invoke(args: Array<out Any?>) = MutableVector((args[0] as Number).toInt(), args.getOrNull(1))
}
